store/src/main/java/com/dropbox/android/external/store4/Fetcher.kt

```diff
@@ -0,0 +1,66 @@
+package com.dropbox.android.external.store4
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.map
+import java.util.concurrent.CancellationException
+
+interface Fetcher<Key, Output> {
+    suspend operator fun invoke(key: Key): Flow<FetcherResponse<Output>>
+
+    companion object {
+        fun <Key, Output> fromValueFetcher(
+            doFetch: (key: Key) -> Flow<Output>
+        ): Fetcher<Key, Output> = FlowingValueFetcher(doFetch)
+
+        fun <Key, Output> fromNonFlowingValueFetcher(
+            doFetch: suspend (key: Key) -> Output
+        ): Fetcher<Key, Output> = NonFlowingValueFetcher(doFetch)
+
+        fun <Key, Output> fromNonFlowingFetcher(
+            doFetch: suspend (key: Key) -> FetcherResponse<Output>
+        ): Fetcher<Key, Output> = NonFlowingFetcher(doFetch)
+    }
+}
```

Comment:
This would simplify the builder indeed (taking the number of builder create function from 8 to 2 in exchange for adding 4 entry points for the fetcher). I would add some builder extension functions to make the common case easy though.

I would like to keep the fetch and transform as separate functions though. The reasons are:
1) The user already needs to transform their internal library results to FetcherResult so it's no additional burden
2) we can enforce that the transform is not a suspend function (it should be quick, the heavy operations should be in the fetcher)
3) less code to maintain (but we can add an extension function if we want the other api back)

How about something like this?
https://gist.github.com/eyalgu/3d03420f72bc0abed76f88be111ce6d2

Reply:
this looks good to me

Reply:
Hmm i'm not sure how much we benefit from `doUnwrap` being explicit. It is all part of user's code so whether is a suspend or not seems unrelated (re: reason 2)? Maybe i'm not fully getting what you mean there.
In fact, i think it would be more annoyance in terms of API if they needed to provide two separate functions for something they could be just handling within their function.